util/test/test_dftool_ganache.py

```diff
@@ -50,6 +51,21 @@ def test_calc(tmp_path):
     assert os.path.exists(rewards_csv)
 
 
+@enforce_types
+@pytest.mark.skip(reason="Requires predictoor support in subgraph")
+def test_predictoor_data(tmp_path):
+    CSV_DIR = str(tmp_path)
+    ST = 0
+    FIN = "latest"
+
+    cmd = f"./dftool predictoor_data {CSV_DIR} {ST} {FIN} {CHAINID}"
+    os.system(cmd)
+
+    # test result
+    predictoor_data_csv = csvs.predictoorDataFilename(CSV_DIR, CHAINID)
+    assert os.path.exists(predictoor_data_csv)
+
+
```

Comment:
You can improve this test:
- Don't aim to test all the functionality end-to-end. Including: don't test subgraph here, you can do that in the specific dedicated unit test.
- Rather, change dftool_module.py::do_predictoor_data to create fake data when it detects that it's in a test. "Design for test (DFT)"; it's a common technique. Then you can also test whether it handles arguments & optional arguments well too.
- I do this already for testing challenge data. Check out [test_dftool_challenge.py ](https://github.com/oceanprotocol/df-py/blob/main/util/challenge/test/test_dftool_challenge.py) and [dftool_module::do_challenge_data](https://github.com/oceanprotocol/df-py/blob/fb0b19d3971c201fdbd815328593368422397979/util/dftool_module.py#L399). See:
```python
    if judge.DFTOOL_TEST_FAKE_CSVDIR in CSV_DIR:
        challenge_data = judge.DFTOOL_TEST_FAKE_CHALLENGE_DATA

    else:  # main path
...
```

Reply:
The thing is, mocks don't work that well when coupled with command line. So they don't record coverage and we can't patch command lines. Furthermore, this test is exactly the one that fails when ran out of order, during my experiments with modularising. Which is me and Berkay discussed a strategy to improve code coverage (see #624) where we could have a 2-pronged approach:

- test the functionality of `do_<something>` functions through pytest
- test the calling of `do_<something>` functions through df_module CLI separately

this way we can use patches more freely and make sure the correct coverage is picked up and mocks work when needed.

Reply:
I agree with the current approach here. 👍